wpiutil/src/main/java/edu/wpi/first/util/struct/DynamicStruct.java

```diff
@@ -390,19 +390,62 @@ public String getStringField(StructFieldDescriptor field) {
     }
     byte[] bytes = new byte[field.m_arraySize];
     m_data.position(field.m_offset).get(bytes, 0, field.m_arraySize);
-    return new String(bytes, StandardCharsets.UTF_8);
+    // Find last non zero character
+    int stringLength = bytes.length;
+    for (; stringLength > 0; stringLength--) {
+      if (bytes[stringLength - 1] != 0) {
+        break;
+      }
+    }
+    // If string is all zeroes, its empty and return an empty string.
+    if (stringLength == 0) {
+      return "";
+    }
+    // Check if the end of the string is in the middle of a continuation byte or not.
+    if ((bytes[stringLength - 1] & 0x80) != 0) {
+      // This is a UTF8 continuation byte. Make sure its valid.
+      // Walk back until initial byte is found
+      int utf8StartByte = stringLength;
+      for (; utf8StartByte > 0; utf8StartByte--) {
+        if ((bytes[utf8StartByte - 1] & 0x40) != 0) {
+          // Having 2nd bit set means start byte
+          break;
+        }
+      }
+      if (utf8StartByte == 0) {
+        // This case means string only contains continuation bytes
+        return "";
+      }
+      utf8StartByte--;
+      // Check if its a 2, 3, or 4 byte
+      byte checkByte = bytes[utf8StartByte];
+      if ((checkByte & 0xE0) == 0xC0 && utf8StartByte != stringLength - 2) {
+        // 2 byte, need 1 more byte
+        stringLength = utf8StartByte;
+      } else if ((checkByte & 0xF0) == 0xE0 && utf8StartByte != stringLength - 3) {
+        // 3 byte, need 2 more bytes
+        stringLength = utf8StartByte;
+      } else if ((checkByte & 0xF8) == 0xF0 && utf8StartByte != stringLength - 4) {
+        // 4 byte, need 3 more bytes
+        stringLength = utf8StartByte;
+      }
```

Comment:
Why isn't this like the C++ version?
Because the first components of these if-else-if conditions (`(checkByte & 0xE0) == 0xC0`, `(checkByte & 0xF0) == 0xE0`, and `(checkByte & 0xF8) == 0xF0`) are mutually exclusive, changing the `&&` to nested ifs (like in the current C++ version) would allow less checks if the encoding is valid. The performance difference is marginal at best, but I'd like for both versions to match as much as possible, and if one's marginally better, might as well go with that one. (If you don't want to change it then that's fine too)

Reply:
You're right, the nested ifs here would be slightly better. Checkstyle wanted this, but I'll change back to the nested ifs and suppress checkstyle.